src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs

```diff
@@ -33,9 +34,11 @@ public object CreateController(ActionContext actionContext)
                     nameof(actionContext));
             }
 
-            var controller = _typeActivator.CreateInstance(
-                _serviceProvider,
-                actionDescriptor.ControllerTypeInfo.AsType());
+            var controllerType = actionDescriptor.ControllerTypeInfo.AsType();
+
+            var fact = _controllerCache.GetOrAdd(controllerType,
+                ActivatorUtilities.CreateFactory(controllerType, Type.EmptyTypes));
+            var controller = fact(_serviceProvider, null);
```

Comment:
This code isn't correct, it's not really caching anything for you.

```
var fact = _controllerCache.GetOrAdd(controllerType, ActivatorUtilities.CreateFactory(controllerType,     Type.EmptyTypes));
```

The reason is that both arguments to `GetOrAdd` must be evaluated before calling the function. That means that `CreateFactory` is going to run every call.

Here's what you can do instead.

```
var fact = _controllerCache.GetOrAdd(
    controllerType, 
    (t) => ActivatorUtilities.CreateFactory(t, Type.EmptyTypes));
```

Now you're allocating a lambda, but you'll only call `CreateFactory` once. 

But wait, there's MORE

You can do even better because the lambda is stateless.

```
private static readonly Func<Type, ObjectFactory> CreateFactory = (t) => ActivatorUtilities.CreateFactory(
    t, 
    Type.EmptyTypes)

...
...
var fact = _controllerCache.GetOrAdd(controllerType, CreateFactory);
```

Now you're guaranteed that the lambda stays static (no allocations required).